{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import varseek as vk\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import json\n",
    "import scipy.sparse as sp\n",
    "from sklearn.metrics import adjusted_rand_score\n",
    "from anndata import concat\n",
    "\n",
    "from varseek.constants import entex_to_ccle_map\n",
    "from varseek.utils import plot_items_descending_order, plot_scree, plot_loading_contributions, find_resolution_for_target_clusters, plot_contingency_table, plot_knn_tissue_frequencies, plot_ascending_bar_plot_of_cluster_distances, plot_jaccard_bar_plot, create_mutated_gene_count_matrix_from_mutation_count_matrix, safe_literal_eval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "threads = 32\n",
    "k = 59\n",
    "minimum_count_filter = 2  # anything below this will be set to 0\n",
    "split_reads_by_Ns_and_low_quality_bases = False\n",
    "use_binary_matrix = False\n",
    "matrix_type = \"mutation\"  # mutation or gene\n",
    "sum_strategy = \"total_reads\"  # only if matrix_type is gene\n",
    "merge_strategy = \"all\"  # only if matrix_type is gene\n",
    "do_cpm_normalization = False\n",
    "mutation_index = \"/home/jrich/data/varseek_data_fresh/vk_build_pipeline_t2t_nov16/mutation_reference.idx\"\n",
    "t2g_vk = \"/home/jrich/data/varseek_data_fresh/vk_build_pipeline_t2t_nov16/t2g_filtered.txt\"\n",
    "\n",
    "standard_index = \"/home/jrich/data/varseek_data/reference/T2T/GCF_009914755.1/kb_index/index.idx\"\n",
    "standard_t2g = \"/home/jrich/data/varseek_data/reference/T2T/GCF_009914755.1/kb_index/t2g.txt\"\n",
    "\n",
    "mutation_metadata_df_path = \"/home/jrich/data/varseek_data_fresh/vk_build_pipeline_t2t_nov16/variants_updated_filtered.csv\"\n",
    "mutation_metadata_df_columns = [\"vcrs_id\", \"gene_name\", \"header_with_gene_name\"]  #* change usecols as desired\n",
    "\n",
    "# Base directory\n",
    "base_dir = \"/home/jrich/data/varseek_data/sequencing/bulk/entex\"\n",
    "entex_csv = \"/home/jrich/data/varseek_data/sequencing/bulk/entex_lung/entex_df.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_out_path = f\"{base_dir}/adata_combined_ccle_rnaseq_{matrix_type}.h5ad\"\n",
    "adata_grouped_tissue_out_path = f\"{base_dir}/adata_combined_ccle_rnaseq_{matrix_type}_grouped_tissue.h5ad\"\n",
    "\n",
    "entex_df = pd.read_csv(entex_csv)\n",
    "\n",
    "entex_df['ccle_primary_disease'] = entex_df['tissue'].map(entex_to_ccle_map).fillna(np.nan)\n",
    "\n",
    "mutation_metadata_df = pd.read_csv(mutation_metadata_df_path, usecols = mutation_metadata_df_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_list = []\n",
    "# Loop through all tissues\n",
    "for tissue_dir in os.listdir(base_dir):\n",
    "    tissue_path = os.path.join(base_dir, tissue_dir)\n",
    "    \n",
    "    # Skip if not a directory\n",
    "    if not os.path.isdir(tissue_path):\n",
    "        continue\n",
    "\n",
    "    # Loop through all samples in the tissue\n",
    "    for sample_dir in os.listdir(tissue_path):\n",
    "        sample_path = os.path.join(tissue_path, sample_dir)\n",
    "        \n",
    "        # Skip if not a directory\n",
    "        if not os.path.isdir(sample_path):\n",
    "            continue\n",
    "\n",
    "        # Paths for pair1 and pair2\n",
    "        pair1_path = os.path.join(sample_path, \"pair1\")\n",
    "        pair2_path = os.path.join(sample_path, \"pair2\")\n",
    "        \n",
    "        # Ensure both directories exist\n",
    "        if os.path.isdir(pair1_path) and os.path.isdir(pair2_path):\n",
    "            # Get the full paths of the FASTQ files (assuming only one file in each folder)\n",
    "            fastq_files = [os.path.join(pair1_path, f) for f in os.listdir(pair1_path) if f.endswith(\".fastq.gz\")]\n",
    "            fastq_files_pair2 = [os.path.join(pair2_path, f) for f in os.listdir(pair2_path) if f.endswith(\".fastq.gz\")]\n",
    "            fastq_files.extend(fastq_files_pair2)\n",
    "\n",
    "            # TODO: replace with vk count\n",
    "            rnaseq_fastq_files_list_dict = vk.fastqpp(\n",
    "                fastq_files,\n",
    "                trim_edges_off_reads=True,\n",
    "                run_fastqc=False,\n",
    "                split_reads_by_Ns_and_low_quality_bases=split_reads_by_Ns_and_low_quality_bases,\n",
    "                parity=\"paired\",\n",
    "                fastqc_out_dir=\".\",\n",
    "                minimum_base_quality_trim_reads=13,\n",
    "                qualified_quality_phred=0,\n",
    "                unqualified_percent_limit=100,\n",
    "                n_base_limit=None,\n",
    "                minimum_length=k,\n",
    "                minimum_base_quality_replace_with_N=13,\n",
    "                fastp=\"fastp\",\n",
    "                seqtk=\"seqtk\",\n",
    "                delete_intermediate_files=True,\n",
    "            )\n",
    "\n",
    "            kb_count_out_standard = f\"{sample_path}/kb_count_standard\"\n",
    "            kb_count_out_mutant = f\"{sample_path}/kb_count_mutant\"\n",
    "\n",
    "            if not os.path.exists(kb_count_out_standard) or len(os.listdir(kb_count_out_standard)) == 0:\n",
    "                kb_count_command = [\n",
    "                    \"kb\",\n",
    "                    \"count\",\n",
    "                    \"-t\",\n",
    "                    str(threads),\n",
    "                    \"-k\",\n",
    "                    str(k),\n",
    "                    \"-i\",\n",
    "                    mutation_index,\n",
    "                    \"-g\",\n",
    "                    t2g_vk,\n",
    "                    \"-x\",\n",
    "                    \"bulk\",\n",
    "                    \"--num\",\n",
    "                    \"--h5ad\",\n",
    "                    \"--parity\",\n",
    "                    \"single\",\n",
    "                    \"--strand\",\n",
    "                    \"unstranded\",\n",
    "                    \"-o\",\n",
    "                    kb_count_out_mutant,\n",
    "                ] + rnaseq_fastq_files_list_dict[\"final\"]\n",
    "                \n",
    "                subprocess.run(kb_count_command, check=True)\n",
    "\n",
    "            if not os.path.exists(kb_count_out_standard) or len(os.listdir(kb_count_out_standard)) == 0:\n",
    "                kb_count_standard_index_command = [\n",
    "                    \"kb\",\n",
    "                    \"count\",\n",
    "                    \"-t\",\n",
    "                    str(threads),\n",
    "                    \"-k\",\n",
    "                    str(k),\n",
    "                    \"-i\",\n",
    "                    standard_index,\n",
    "                    \"-g\",\n",
    "                    standard_t2g,\n",
    "                    \"-x\",\n",
    "                    \"bulk\",\n",
    "                    \"--h5ad\",\n",
    "                    \"--parity\",\n",
    "                    \"paired\",\n",
    "                    \"--strand\",\n",
    "                    \"single\",\n",
    "                    \"-o\",\n",
    "                    kb_count_out_standard,\n",
    "                ] + rnaseq_fastq_files_list_dict[\"trimmed\"]\n",
    "                \n",
    "                subprocess.run(kb_count_standard_index_command, check=True)\n",
    "\n",
    "        adata_path = f\"{kb_count_out_mutant}/counts_unfiltered/adata.h5ad\"\n",
    "        adata_output_path = f\"{kb_count_out_mutant}/counts_unfiltered/adata_cleaned.h5ad\"\n",
    "\n",
    "        adata_path_normal_genome = f\"{kb_count_out_standard}/counts_unfiltered/adata.h5ad\"\n",
    "\n",
    "        adata_mutation = sc.read_h5ad(adata_path)\n",
    "        adata_mutation.obs_names = [sample_dir]\n",
    "        adata_mutation.obs['experiment_id'] = sample_dir\n",
    "        \n",
    "        adata_mutation = vk.varseek_clean.clean(adata_mutation, adata_output_path = adata_output_path, minimum_count_filter = minimum_count_filter, use_binary_matrix = use_binary_matrix, technology = \"bulk\", do_cpm_normalization = do_cpm_normalization, split_reads_by_Ns_and_low_quality_bases = split_reads_by_Ns_and_low_quality_bases)\n",
    "\n",
    "        if 'vcrs_count' in adata_mutation.var.columns:\n",
    "            adata_mutation.var.drop('vcrs_count', axis=1, inplace=True)\n",
    "\n",
    "        adata_list.append(adata_mutation)\n",
    "\n",
    "adata = concat(adata_list, join='outer')\n",
    "adata.obs = adata.obs.merge(\n",
    "    entex_df,\n",
    "    on='experiment_id', \n",
    "    how='left'\n",
    ")\n",
    "\n",
    "# prevents the error \"Donâ€™t call _normalize_index with non-categorical/string names\" when trying adata = adata[:, nonzero_gene_mask]\n",
    "adata.obs.index = adata.obs.index.astype(str)\n",
    "\n",
    "if use_binary_matrix:\n",
    "    adata.X = (adata.X > 0).astype(int)\n",
    "\n",
    "adata.var['vcrs_id'] = adata.var.index\n",
    "original_var_names = adata.var_names.copy()\n",
    "# merge adata.var with mutation_metadata_df by vcrs_id\n",
    "adata.var = adata.var.merge(mutation_metadata_df, on='vcrs_id', how='left')\n",
    "adata.var_names = original_var_names\n",
    "\n",
    "adata.var[\"vcrs_count\"] = adata.X.sum(axis=0).A1 if hasattr(adata.X, \"A1\") else np.asarray(adata.X.sum(axis=0)).flatten()\n",
    "\n",
    "adata.var = adata.var.copy()  # avoids warning\n",
    "adata.var[\"gene_name\"] = adata.var[\"gene_name\"].astype('object')  # converts from category to object\n",
    "adata.var[\"gene_name\"] = adata.var[\"gene_name\"].apply(safe_literal_eval)  # converts list-like string to list\n",
    "adata.var[\"gene_name_set\"] = adata.var['gene_name'].apply(lambda x: sorted(set(x)))\n",
    "adata.var['gene_name_set_string'] = adata.var['gene_name_set'].apply(lambda x: ';'.join(x))  # converts list to semi-colon joined string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if matrix_type == \"gene\":\n",
    "    adata_mutation = adata.copy()\n",
    "    adata = create_mutated_gene_count_matrix_from_mutation_count_matrix(adata, sum_strategy = sum_strategy, merge_strategy = merge_strategy, use_binary_matrix=use_binary_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var['gene_name'] = adata.var['gene_name'].apply(str)  # Convert gene_name to string to avoid error when saving adata\n",
    "adata.var['gene_name_set'] = adata.var['gene_name_set'].apply(str)  # Convert gene_name to string to avoid error when saving adata\n",
    "\n",
    "# group by tissue\n",
    "#!!! DO THIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do analysis\n",
    "adata_ccle_grouped_path = \"/home/jrich/data/varseek_data/sequencing/bulk/ccle/adata_combined_ccle_rnaseq_mutation_grouped_tissue.h5ad\"\n",
    "adata_ccle = sc.read_h5ad(adata_ccle_grouped_path)\n",
    "\n",
    "#!!! DO THIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(adata_out_path)\n",
    "adata_tissue_grouped.write(adata_grouped_tissue_out_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "varseek2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
