{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8a4b538-0583-45e9-91d1-eec859bae4c1",
   "metadata": {},
   "source": [
    "# Extract cancer variant reads for alignment visualization\n",
    "\n",
    "#### Requirements: kb, samtools, and bowtie2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e3f9a920-1ebd-4cc2-94eb-f843fa9d98e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to COSMIC vk build data frame\n",
    "cosmic_gget_mut_df = \"/home/jrich/data/varseek_data/reference/cosmic/CancerMutationCensus_AllData_Tsv_v100_GRCh37/CancerMutationCensus_AllData_v100_GRCh37_mutation_workflow_with_cdna.csv\"\n",
    "vcrs_fa_path = \"/home/jrich/data/varseek_data_fresh/vk_build_pipeline_grch37_ensembl93/vcrs_filtered.fa\"\n",
    "base_out = \"/home/jrich/data/varseek_data_fresh/alignment_visualization\"\n",
    "index = \"/home/jrich/data/varseek_data_fresh/vk_build_pipeline_grch37_ensembl93/mutation_reference.idx\"\n",
    "t2g = \"/home/jrich/data/varseek_data_fresh/vk_build_pipeline_grch37_ensembl93/t2g_filtered.txt\"\n",
    "k = \"59\"\n",
    "\n",
    "# RNAseq of OE21_OESOPHAGUS\n",
    "fastq1 = \"/home/COH_data/cancer_variants_stuff/cancer_cell_lines/SRR8615227_1.fastq.gz\"\n",
    "fastq2 = \"/home/COH_data/cancer_variants_stuff/cancer_cell_lines/SRR8615227_2.fastq.gz\"\n",
    "\n",
    "# Path to where shortened copy of fastq1 will be saved (only used to test this notebook)\n",
    "fastq1_short = f\"{base_out}/SRR8615227_1_short.fastq.gz\"\n",
    "fastq2_short = f\"{base_out}/SRR8615227_2_short.fastq.gz\"\n",
    "\n",
    "# Directory to save bowtie index in\n",
    "b_index = f\"{base_out}/bowtie_indices\"\n",
    "\n",
    "# Directory to save kb count output in\n",
    "out = f\"{base_out}/SRR8615227\"\n",
    "\n",
    "# Directory to save kb extract output in\n",
    "extract_out = f\"{base_out}/SRR8615227/extract_out\"\n",
    "\n",
    "bowtie2 = \"/home/jrich/opt/bowtie2-2.5.4/bowtie2-2.5.4-linux-x86_64/bowtie2\"\n",
    "bowtie_out = f\"{base_out}/bowtie_alignments\"\n",
    "samtools = \"/home/jrich/miniconda3/envs/kvar/bin/samtools\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "753595a2-e288-4133-856e-dd6df1076948",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyfastx\n",
    "# pd.set_option('display.max_columns', None)\n",
    "\n",
    "threads = 16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d408b021",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create new file keeping only first X reads (to test this workflow)\n",
    "!zcat $fastq1 | head -8000000 > $fastq1_short\n",
    "# !zcat $fastq2 | head -8000000 > $fastq2_short"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "860c74b1-f920-430a-bf39-b3f9e03cbfe9",
   "metadata": {},
   "source": [
    "### Align test dataset to cancer index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e7a389fc-8973-4421-a67f-85f358b0719d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2024-11-05 16:20:11,794]    INFO [count] Using index /home/jrich/data/varseek_data_fresh/vk_build_pipeline/mutation_reference.idx to generate BUS file to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227 from\n",
      "[2024-11-05 16:20:11,794]    INFO [count]         /home/COH_data/cancer_variants_stuff/cancer_cell_lines/SRR8615227_1.fastq.gz\n",
      "[2024-11-05 16:20:11,794]    INFO [count]         /home/COH_data/cancer_variants_stuff/cancer_cell_lines/SRR8615227_2.fastq.gz\n",
      "[2024-11-05 16:27:11,115]    INFO [count] Sorting BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/tmp/output.s.bus\n",
      "[2024-11-05 16:27:14,825]    INFO [count] Inspecting BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/tmp/output.s.bus\n",
      "[2024-11-05 16:27:15,928]    INFO [count] Generating count matrix /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/counts_unfiltered/cells_x_genes from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/tmp/output.s.bus\n",
      "[2024-11-05 16:27:47,233]    INFO [count] Writing gene names to file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/counts_unfiltered/cells_x_genes.genes.names.txt\n",
      "[2024-11-05 16:27:53,323] WARNING [count] 5259437 gene IDs do not have corresponding valid gene names. These genes will use their gene IDs instead.\n",
      "[2024-11-05 16:27:54,249]    INFO [count] Reading matrix /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/counts_unfiltered/cells_x_genes.mtx\n",
      "[2024-11-05 16:28:08,818]    INFO [count] Writing matrix to h5ad /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/counts_unfiltered/adata.h5ad\n",
      "CPU times: user 10.2 s, sys: 1.79 s, total: 12 s\n",
      "Wall time: 8min 5s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# running only 1 fastq and doing single-end mode because kb extract only takes in 1 fastq; also, using short fastq instead of full one so that I extract the top reads within this fastq, not overall (both of these things differ from original run - see git history before 11/11 for original)\n",
    "\n",
    "!kb count \\\n",
    "    -t $threads \\\n",
    "    -i $index \\\n",
    "    -g $t2g \\\n",
    "    -x bulk \\\n",
    "    -k $k \\\n",
    "    --h5ad \\\n",
    "    --parity single \\\n",
    "    -o $out \\\n",
    "    $fastq1_short"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee383af2-99e5-49b3-bcbd-429200047525",
   "metadata": {},
   "source": [
    "### Pull out reads for top 10 mutants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d68cd632-8354-4faf-b1da-86e72f993f4e",
   "metadata": {},
   "source": [
    "Identify top 10 mutants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "21780862-fa1b-4b05-ad9b-e15b2b416820",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1 × 5259437"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = anndata.read_h5ad(f\"{out}/counts_unfiltered/adata.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "615b0e87-2868-4f7d-8db7-91c93c3d025a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['seq_0688768', 'seq_3311555', 'seq_5257480', 'seq_5188714',\n",
       "       'seq_1338451', 'seq_5257481', 'seq_1668855', 'seq_1668852',\n",
       "       'seq_4327812', 'seq_1668681'],\n",
       "      dtype='object', name='gene_id')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top10 = adata.var.index[np.argsort(np.array(adata.X.todense())[0])][::-1][:10]\n",
    "top10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c74466a5-2867-4332-b270-1e085a845fd5",
   "metadata": {},
   "source": [
    "Show metadata for top 10 mutations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b9e6c27c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_312178/363827191.py:1: DtypeWarning: Columns (4,5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  cosmic_df = pd.read_csv(cosmic_gget_mut_df)\n"
     ]
    }
   ],
   "source": [
    "cosmic_df = pd.read_csv(cosmic_gget_mut_df)\n",
    "cosmic_df[\"seq_ID:mutation\"] = cosmic_df[\"seq_ID\"] + \":\" + cosmic_df[\"mutation\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7e7b11cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = \"/home/jrich/data/varseek_data_fresh/vk_build_pipeline_grch37_ensembl93/variants_updated_filtered.csv\"\n",
    "columns_to_load = [\"header_cds\", \"vcrs_id\"]\n",
    "\n",
    "metadata_df = pd.read_csv(file_path, usecols=columns_to_load)\n",
    "\n",
    "# Function to process each entry\n",
    "def process_header_cds(s):\n",
    "    s = s.strip(\"[]'\")  # Remove brackets and quotes\n",
    "    entries = s.split(\", \")  # Split by comma and space\n",
    "    return ';'.join(entries)  # Join with semicolon\n",
    "\n",
    "# Apply the function to the column\n",
    "metadata_df['header_cds'] = metadata_df['header_cds'].apply(process_header_cds).astype(str)\n",
    "\n",
    "id_to_cds_header_dict = dict(zip(metadata_df['vcrs_id'], metadata_df['header_cds']))\n",
    "top10_headers = [id_to_cds_header_dict[x] for x in top10]\n",
    "\n",
    "# Convert the dictionary to a DataFrame\n",
    "id_to_cds_header_df = pd.DataFrame(list(id_to_cds_header_dict.items()), columns=['vcrs_id', 'seq_ID:mutation'])\n",
    "\n",
    "# Merge with cosmic_df\n",
    "cosmic_df = cosmic_df.merge(id_to_cds_header_df, on='seq_ID:mutation', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "73d45184",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_ID</th>\n",
       "      <th>mutation_cds</th>\n",
       "      <th>mutation_aa</th>\n",
       "      <th>GENOMIC_MUTATION_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>mutation_genome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mutation_id</th>\n",
       "      <th>mutation</th>\n",
       "      <th>seq_ID:mutation</th>\n",
       "      <th>vcrs_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>689965</th>\n",
       "      <td>ENST00000362079</td>\n",
       "      <td>c.174G&gt;A</td>\n",
       "      <td>p.W58=</td>\n",
       "      <td>COSV104668806</td>\n",
       "      <td>MT</td>\n",
       "      <td>+</td>\n",
       "      <td>g.9380G&gt;A</td>\n",
       "      <td>MT-CO3</td>\n",
       "      <td>37406576</td>\n",
       "      <td>c.174G&gt;A</td>\n",
       "      <td>ENST00000362079:c.174G&gt;A</td>\n",
       "      <td>seq_0688768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1341346</th>\n",
       "      <td>ENST00000331825</td>\n",
       "      <td>c.163T&gt;C</td>\n",
       "      <td>p.L55=</td>\n",
       "      <td>COSV53169473</td>\n",
       "      <td>19</td>\n",
       "      <td>+</td>\n",
       "      <td>g.49469087T&gt;C</td>\n",
       "      <td>FTL</td>\n",
       "      <td>26232420</td>\n",
       "      <td>c.370T&gt;C</td>\n",
       "      <td>ENST00000331825:c.163T&gt;C</td>\n",
       "      <td>seq_1338451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1673880</th>\n",
       "      <td>ENST00000252242</td>\n",
       "      <td>c.351C&gt;T</td>\n",
       "      <td>p.L117=</td>\n",
       "      <td>COSV52862629</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>g.52913730G&gt;A</td>\n",
       "      <td>KRT5</td>\n",
       "      <td>22187999</td>\n",
       "      <td>c.742C&gt;T</td>\n",
       "      <td>ENST00000252242:c.351C&gt;T</td>\n",
       "      <td>seq_1668681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1674052</th>\n",
       "      <td>ENST00000252242</td>\n",
       "      <td>c.732G&gt;A</td>\n",
       "      <td>p.L244=</td>\n",
       "      <td>COSV107254094</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>g.52912768C&gt;T</td>\n",
       "      <td>KRT5</td>\n",
       "      <td>22187988</td>\n",
       "      <td>c.1123G&gt;A</td>\n",
       "      <td>ENST00000252242:c.732G&gt;A</td>\n",
       "      <td>seq_1668852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1674055</th>\n",
       "      <td>ENST00000252242</td>\n",
       "      <td>c.1065A&gt;C</td>\n",
       "      <td>p.T355=</td>\n",
       "      <td>COSV52861331</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>g.52911401T&gt;G</td>\n",
       "      <td>KRT5</td>\n",
       "      <td>22187981</td>\n",
       "      <td>c.1456A&gt;C</td>\n",
       "      <td>ENST00000252242:c.1065A&gt;C</td>\n",
       "      <td>seq_1668855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3331623</th>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>c.609G&gt;A</td>\n",
       "      <td>p.G203=</td>\n",
       "      <td>COSV62293499</td>\n",
       "      <td>MT</td>\n",
       "      <td>+</td>\n",
       "      <td>g.3915G&gt;A</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>38370056</td>\n",
       "      <td>c.609G&gt;A</td>\n",
       "      <td>ENST00000361390:c.609G&gt;A</td>\n",
       "      <td>seq_3311555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4357198</th>\n",
       "      <td>ENST00000330722</td>\n",
       "      <td>c.819T&gt;C</td>\n",
       "      <td>p.D273=</td>\n",
       "      <td>COSV58104299</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>g.52884735A&gt;G</td>\n",
       "      <td>KRT6A</td>\n",
       "      <td>29208107</td>\n",
       "      <td>c.888T&gt;C</td>\n",
       "      <td>ENST00000330722:c.819T&gt;C</td>\n",
       "      <td>seq_4327812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5229642</th>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>c.189dup</td>\n",
       "      <td>p.A64Rfs*47</td>\n",
       "      <td>COSV106517221</td>\n",
       "      <td>MT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.3495_3496dup</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>38371054</td>\n",
       "      <td>c.189dup</td>\n",
       "      <td>ENST00000361390:c.189dup</td>\n",
       "      <td>seq_5188714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298953</th>\n",
       "      <td>ENST00000361624</td>\n",
       "      <td>c.516del</td>\n",
       "      <td>p.K172Nfs*?</td>\n",
       "      <td>COSV107450106</td>\n",
       "      <td>MT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.6419del</td>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>35129715</td>\n",
       "      <td>c.516del</td>\n",
       "      <td>ENST00000361624:c.516del</td>\n",
       "      <td>seq_5257480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298954</th>\n",
       "      <td>ENST00000361624</td>\n",
       "      <td>c.795dup</td>\n",
       "      <td>p.E266Rfs*?</td>\n",
       "      <td>COSV104419744</td>\n",
       "      <td>MT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.6698_6699dup</td>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>35129625</td>\n",
       "      <td>c.795dup</td>\n",
       "      <td>ENST00000361624:c.795dup</td>\n",
       "      <td>seq_5257481</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  seq_ID mutation_cds  mutation_aa GENOMIC_MUTATION_ID  \\\n",
       "689965   ENST00000362079     c.174G>A       p.W58=       COSV104668806   \n",
       "1341346  ENST00000331825     c.163T>C       p.L55=        COSV53169473   \n",
       "1673880  ENST00000252242     c.351C>T      p.L117=        COSV52862629   \n",
       "1674052  ENST00000252242     c.732G>A      p.L244=       COSV107254094   \n",
       "1674055  ENST00000252242    c.1065A>C      p.T355=        COSV52861331   \n",
       "3331623  ENST00000361390     c.609G>A      p.G203=        COSV62293499   \n",
       "4357198  ENST00000330722     c.819T>C      p.D273=        COSV58104299   \n",
       "5229642  ENST00000361390     c.189dup  p.A64Rfs*47       COSV106517221   \n",
       "5298953  ENST00000361624     c.516del  p.K172Nfs*?       COSV107450106   \n",
       "5298954  ENST00000361624     c.795dup  p.E266Rfs*?       COSV104419744   \n",
       "\n",
       "        chromosome strand mutation_genome gene_name  mutation_id   mutation  \\\n",
       "689965          MT      +       g.9380G>A    MT-CO3     37406576   c.174G>A   \n",
       "1341346         19      +   g.49469087T>C       FTL     26232420   c.370T>C   \n",
       "1673880         12      -   g.52913730G>A      KRT5     22187999   c.742C>T   \n",
       "1674052         12      -   g.52912768C>T      KRT5     22187988  c.1123G>A   \n",
       "1674055         12      -   g.52911401T>G      KRT5     22187981  c.1456A>C   \n",
       "3331623         MT      +       g.3915G>A    MT-ND1     38370056   c.609G>A   \n",
       "4357198         12      -   g.52884735A>G     KRT6A     29208107   c.888T>C   \n",
       "5229642         MT    NaN  g.3495_3496dup    MT-ND1     38371054   c.189dup   \n",
       "5298953         MT    NaN       g.6419del    MT-CO1     35129715   c.516del   \n",
       "5298954         MT    NaN  g.6698_6699dup    MT-CO1     35129625   c.795dup   \n",
       "\n",
       "                   seq_ID:mutation      vcrs_id  \n",
       "689965    ENST00000362079:c.174G>A  seq_0688768  \n",
       "1341346   ENST00000331825:c.163T>C  seq_1338451  \n",
       "1673880   ENST00000252242:c.351C>T  seq_1668681  \n",
       "1674052   ENST00000252242:c.732G>A  seq_1668852  \n",
       "1674055  ENST00000252242:c.1065A>C  seq_1668855  \n",
       "3331623   ENST00000361390:c.609G>A  seq_3311555  \n",
       "4357198   ENST00000330722:c.819T>C  seq_4327812  \n",
       "5229642   ENST00000361390:c.189dup  seq_5188714  \n",
       "5298953   ENST00000361624:c.516del  seq_5257480  \n",
       "5298954   ENST00000361624:c.795dup  seq_5257481  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosmic_df[cosmic_df[\"seq_ID:mutation\"].isin(top10_headers)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d0be8b35-9a12-4a97-a514-84c6c72f9a0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2024-11-05 17:30:34,496]    INFO [extract] Performing alignment using kallisto...\n",
      "[2024-11-05 17:30:34,497]    INFO [extract] Using index /home/jrich/data/varseek_data_fresh/vk_build_pipeline/mutation_reference.idx to generate BUS file to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp from\n",
      "[2024-11-05 17:30:34,497]    INFO [extract]         /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227_1_short.fastq.gz\n",
      "[2024-11-05 17:31:18,044]    INFO [extract] Alignment complete. Beginning extraction of reads using bustools...\n",
      "[2024-11-05 18:00:28,579]    INFO [extract] Extracting reads for the following transcript: seq_0688768\n",
      "[2024-11-05 18:00:28,579]    INFO [extract] Capturing records from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_0688768.bus with capture list /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/pull_out_reads_transcript_ids_temp.txt\n",
      "[2024-11-05 18:00:39,305]    INFO [extract] Sorting BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_0688768.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_0688768_sorted.bus\n",
      "[2024-11-05 18:00:42,413]    INFO [extract] Extracting BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_0688768_sorted.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/seq_0688768\n",
      "[2024-11-05 18:00:44,164]    INFO [extract] Extracting reads for the following transcript: seq_3311555\n",
      "[2024-11-05 18:00:44,165]    INFO [extract] Capturing records from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_3311555.bus with capture list /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/pull_out_reads_transcript_ids_temp.txt\n",
      "[2024-11-05 18:00:54,691]    INFO [extract] Sorting BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_3311555.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_3311555_sorted.bus\n",
      "[2024-11-05 18:00:57,900]    INFO [extract] Extracting BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_3311555_sorted.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/seq_3311555\n",
      "[2024-11-05 18:00:59,439]    INFO [extract] Extracting reads for the following transcript: seq_5257480\n",
      "[2024-11-05 18:00:59,439]    INFO [extract] Capturing records from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_5257480.bus with capture list /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/pull_out_reads_transcript_ids_temp.txt\n",
      "[2024-11-05 18:01:08,860]    INFO [extract] Sorting BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_5257480.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_5257480_sorted.bus\n",
      "[2024-11-05 18:01:11,968]    INFO [extract] Extracting BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_5257480_sorted.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/seq_5257480\n",
      "[2024-11-05 18:01:13,593]    INFO [extract] Extracting reads for the following transcript: seq_5188714\n",
      "[2024-11-05 18:01:13,593]    INFO [extract] Capturing records from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_5188714.bus with capture list /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/pull_out_reads_transcript_ids_temp.txt\n",
      "[2024-11-05 18:01:23,416]   ERROR [extract] Extraction of reads unsuccessful for seq_5188714 due to the following error:\n",
      "/home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_5188714.bus has no BUS records\n",
      "[2024-11-05 18:01:23,417]    INFO [extract] Extracting reads for the following transcript: seq_1338451\n",
      "[2024-11-05 18:01:23,417]    INFO [extract] Capturing records from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_1338451.bus with capture list /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/pull_out_reads_transcript_ids_temp.txt\n",
      "[2024-11-05 18:01:31,637]   ERROR [extract] Extraction of reads unsuccessful for seq_1338451 due to the following error:\n",
      "/home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_1338451.bus has no BUS records\n",
      "[2024-11-05 18:01:31,637]    INFO [extract] Extracting reads for the following transcript: seq_5257481\n",
      "[2024-11-05 18:01:31,637]    INFO [extract] Capturing records from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_5257481.bus with capture list /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/pull_out_reads_transcript_ids_temp.txt\n",
      "[2024-11-05 18:01:39,957]    INFO [extract] Sorting BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_5257481.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_5257481_sorted.bus\n",
      "[2024-11-05 18:01:43,065]    INFO [extract] Extracting BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_5257481_sorted.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/seq_5257481\n",
      "[2024-11-05 18:01:44,605]    INFO [extract] Extracting reads for the following transcript: seq_1668855\n",
      "[2024-11-05 18:01:44,605]    INFO [extract] Capturing records from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_1668855.bus with capture list /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/pull_out_reads_transcript_ids_temp.txt\n",
      "[2024-11-05 18:01:52,323]   ERROR [extract] Extraction of reads unsuccessful for seq_1668855 due to the following error:\n",
      "/home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_1668855.bus has no BUS records\n",
      "[2024-11-05 18:01:52,356]    INFO [extract] Extracting reads for the following transcript: seq_1668852\n",
      "[2024-11-05 18:01:52,356]    INFO [extract] Capturing records from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_1668852.bus with capture list /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/pull_out_reads_transcript_ids_temp.txt\n",
      "[2024-11-05 18:02:00,676]   ERROR [extract] Extraction of reads unsuccessful for seq_1668852 due to the following error:\n",
      "/home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_1668852.bus has no BUS records\n",
      "[2024-11-05 18:02:00,714]    INFO [extract] Extracting reads for the following transcript: seq_4327812\n",
      "[2024-11-05 18:02:00,714]    INFO [extract] Capturing records from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_4327812.bus with capture list /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/pull_out_reads_transcript_ids_temp.txt\n",
      "[2024-11-05 18:02:10,135]   ERROR [extract] Extraction of reads unsuccessful for seq_4327812 due to the following error:\n",
      "/home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_4327812.bus has no BUS records\n",
      "[2024-11-05 18:02:10,169]    INFO [extract] Extracting reads for the following transcript: seq_1668681\n",
      "[2024-11-05 18:02:10,170]    INFO [extract] Capturing records from BUS file /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output.bus to /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_1668681.bus with capture list /home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/pull_out_reads_transcript_ids_temp.txt\n",
      "[2024-11-05 18:02:17,887]   ERROR [extract] Extraction of reads unsuccessful for seq_1668681 due to the following error:\n",
      "/home/jrich/data/varseek_data_fresh/alignment_visualization/SRR8615227/extract_out/tmp/output_extracted_seq_1668681.bus has no BUS records\n"
     ]
    }
   ],
   "source": [
    "# TODO: replace this with my bustools df command\n",
    "\n",
    "targets = \" \".join(top10)\n",
    "\n",
    "# Replace fastq1_short with fastq1 to extract all reads\n",
    "!kb extract \\\n",
    "    -t $threads \\\n",
    "    -g $t2g \\\n",
    "    -i $index \\\n",
    "    -k $k \\\n",
    "    -ts $targets \\\n",
    "    -ttype transcript \\\n",
    "    -o $extract_out \\\n",
    "    $fastq1_short"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "142ba588-5875-4394-8c03-4dd6b08f0568",
   "metadata": {},
   "source": [
    "### Align extracted reads with bowtie2 to create bam file for alignment visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edda2df2-d393-4cb3-9560-0f3a4820404d",
   "metadata": {},
   "source": [
    "#### Get/build bowtie2 index for human genome:\n",
    "\n",
    "I opted for downloading the pre-generated bowtie2 index for 'Human / GRCh38 no-alt analysis set' from https://benlangmead.github.io/aws-indexes/bowtie (05/24/2024)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "4d25e1df-29fb-4450-adce-c0be77bbb7c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 3575M  100 3575M    0     0  9690k      0  0:06:17  0:06:17 --:--:--  9.8M75M    2 88.9M    0     0  9107k      0  0:06:42  0:00:09  0:06:33  9.8M35 3575M   35 1256M    0     0  8750k      0  0:06:58  0:02:26  0:04:32  9.7M8 3575M   38 1374M    0     0  8795k      0  0:06:56  0:02:40  0:04:16 9745k3575M   84 3016M    0     0  9532k      0  0:06:24  0:05:23  0:01:01 9462k3575M   91 3260M    0     0  9621k      0  0:06:20  0:05:46  0:00:34 8340kM    0     0  9695k      0  0:06:17  0:06:05  0:00:12 11.1M7M\n",
      "Archive:  GRCh38_noalt_as.zip\n",
      "   creating: GRCh38_noalt_as/\n",
      "  inflating: GRCh38_noalt_as/GRCh38_noalt_as.3.bt2  \n",
      "  inflating: GRCh38_noalt_as/GRCh38_noalt_as.rev.2.bt2  \n",
      "  inflating: GRCh38_noalt_as/GRCh38_noalt_as.rev.1.bt2  \n",
      "  inflating: GRCh38_noalt_as/GRCh38_noalt_as.1.bt2  \n",
      "  inflating: GRCh38_noalt_as/GRCh38_noalt_as.2.bt2  \n",
      "  inflating: GRCh38_noalt_as/GRCh38_noalt_as.4.bt2  \n"
     ]
    }
   ],
   "source": [
    "os.makedirs(b_index, exist_ok=True)\n",
    "if not os.path.exists(f\"{b_index}/GRCh38_noalt_as\"):\n",
    "    !cd $b_index && curl -O https://genome-idx.s3.amazonaws.com/bt/GRCh38_noalt_as.zip && unzip GRCh38_noalt_as.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "487178cb-9825-4c23-9551-baf7a6934f9c",
   "metadata": {},
   "source": [
    "Align pulled out reads to human genome to generate bam files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "28187021-d85a-4bd0-9892-7fd5e20e46e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "seq_5257481\n",
      "7681 reads; of these:\n",
      "  7681 (100.00%) were unpaired; of these:\n",
      "    35 (0.46%) aligned 0 times\n",
      "    7646 (99.54%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "99.54% overall alignment rate\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "\n",
      "seq_5257480\n",
      "19 reads; of these:\n",
      "  19 (100.00%) were unpaired; of these:\n",
      "    1 (5.26%) aligned 0 times\n",
      "    18 (94.74%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "94.74% overall alignment rate\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "\n",
      "seq_0688768\n",
      "2458 reads; of these:\n",
      "  2458 (100.00%) were unpaired; of these:\n",
      "    199 (8.10%) aligned 0 times\n",
      "    2259 (91.90%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "91.90% overall alignment rate\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "\n",
      "seq_3311555\n",
      "196 reads; of these:\n",
      "  196 (100.00%) were unpaired; of these:\n",
      "    4 (2.04%) aligned 0 times\n",
      "    192 (97.96%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "97.96% overall alignment rate\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "CPU times: user 632 ms, sys: 386 ms, total: 1.02 s\n",
      "Wall time: 33.7 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import glob\n",
    "data = extract_out\n",
    "\n",
    "vcrs_fa = pyfastx.Fasta(vcrs_fa_path, build_index=True)\n",
    "\n",
    "os.makedirs(bowtie_out, exist_ok=True)\n",
    "\n",
    "for folder in glob.glob(data + \"/*/\"):\n",
    "    variant = folder.split(\"/\")[-2]\n",
    "    print(\"\\n\" + variant)\n",
    "\n",
    "    outfolder = f\"{bowtie_out}/{variant}\"\n",
    "    !mkdir -p $outfolder\n",
    "    \n",
    "    # Align reads to human ref using bowtie2\n",
    "    variant_sam = variant + \".sam\"\n",
    "    !$bowtie2 \\\n",
    "        --very-sensitive \\\n",
    "        -k 1 \\\n",
    "        -x $b_index/GRCh38_noalt_as/GRCh38_noalt_as \\\n",
    "        -p $threads \\\n",
    "        -q $data/$variant/1.fastq.gz \\\n",
    "        -S $outfolder/$variant_sam\n",
    "\n",
    "    # Convert sam to bam\n",
    "    variant_bam = variant + \".bam\"\n",
    "    !$samtools view \\\n",
    "        -bS -F4 $outfolder/$variant_sam \\\n",
    "        > $outfolder/$variant_bam\n",
    "    \n",
    "    # Sort bam file\n",
    "    variant_bam_sorted = variant + \"_sorted.bam\"\n",
    "    !$samtools sort \\\n",
    "        $outfolder/$variant_bam \\\n",
    "        -o $outfolder/$variant_bam_sorted\n",
    "    \n",
    "    # Create an index for the sorted bam file (creates a .bai file)\n",
    "    !$samtools index $outfolder/$variant_bam_sorted\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # repeat but for vcrs\n",
    "    outfolder_vcrs = f\"{bowtie_out}/{variant}_vcrs\"\n",
    "    !mkdir -p $outfolder_vcrs\n",
    "    variant_reference_sequence = vcrs_fa[variant].seq.strip()\n",
    "\n",
    "    # Align reads to human ref using bowtie2\n",
    "    variant_sam = variant + \".sam\"\n",
    "    !$bowtie2 \\\n",
    "        --very-sensitive \\\n",
    "        -k 1 \\\n",
    "        -x $b_index/GRCh38_noalt_as/GRCh38_noalt_as \\\n",
    "        -p $threads \\\n",
    "        -c $variant_reference_sequence \\\n",
    "        -S $outfolder_vcrs/$variant_sam\n",
    "\n",
    "    # Convert sam to bam\n",
    "    variant_bam = variant + \".bam\"\n",
    "    !$samtools view \\\n",
    "        -bS -F4 $outfolder_vcrs/$variant_sam \\\n",
    "        > $outfolder_vcrs/$variant_bam\n",
    "\n",
    "    # Sort bam file\n",
    "    variant_bam_sorted = variant + \"_sorted_vcrs.bam\"\n",
    "    !$samtools sort \\\n",
    "        $outfolder_vcrs/$variant_bam \\\n",
    "        -o $outfolder_vcrs/$variant_bam_sorted\n",
    "\n",
    "    # Create an index for the sorted bam file (creates a .bai file)\n",
    "    !$samtools index $outfolder_vcrs/$variant_bam_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "443d6d26-7f24-4344-b6e1-90573e49c458",
   "metadata": {},
   "source": [
    "#### Bam files downloaded to local computer with `scp -r laura@dator:/home/laura/projects/cancer_variants/data/bowtie_alignments ./` and visualized using NCBI Genome workbench\n",
    "\n",
    "for me: scp -r jrich@dator:/home/jrich/data/varseek_data_fresh/alignment_visualization/bowtie_alignments ./"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "varseek2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
